Firestore/Swift/Source/Result/DocumentReference+Result.swift

```diff
@@ -0,0 +1,247 @@
+    ///   - source: The `FirestoreSource` where the results should be fetched:
+    ///     - `default`:  Fetch from the server and, if it fails, the cache.
+    ///     - `server`:  Fetch from the server only.
+    ///     - `cache`:  Fetch from the cache only.
+    ///   - completion: The closure to execute on receipt of a result.
+    ///   - result: The result of request. On success it contains the `DocumentSnapshot`, otherwise an `Error`.
+    func getDocument(source: FirestoreSource = .default,
+                     completion: @escaping (_ result: Result<DocumentSnapshot, Error>) -> Void) {
+      getDocument(source: source, completion: mapResultClosure(completion))
+    }
+
+    /// Attaches a listener for this `DocumentReference` events.
+    ///
+    /// - Parameters:
+    ///   - includeMetadataChanges: Indicates if metadata changes  should trigger snapshot events.
+    ///   - listenerHandler: The closure to execute on receipt of a result.
+    ///   - result: The result of request. On success it contains the `QuerySnapshot`, otherwise an `Error`.
+    /// - Returns: The `ListenerRegistration` that can be used to remove this listener.
+    func addSnapshotListener(includeMetadataChanges: Bool = false,
+                             listener: @escaping (_ result: Result<DocumentSnapshot, Error>)
+                               -> Void)
+      -> ListenerRegistration {
+      addSnapshotListener(
+        includeMetadataChanges: includeMetadataChanges,
+        listener: mapResultClosure(listener)
+      )
+    }
+  }
+
+  /// Returns a closure mapped from the a given closure with a `Result` parameter.
+  ///
+  /// - Precondition:
+  ///   Internal return value and error must not both be nil.
+  ///
+  /// - Parameters:
+  ///   - completion: The closure to map.
+  ///   - result: The parameter of the closure to map.
+  /// - Returns: A closure mapped from the given closure.
+  private func mapResultClosure<T>(_ completion: @escaping (_ result: Result<T, Error>) -> Void)
+    -> ((T?, Error?) -> Void) {
+    { value, error in
+      if let value = value {
+        completion(.success(value))
+      } else if let error = error {
+        completion(.failure(error))
+      } else {
+        // preconditionFailure("Internal return value and error must not both be nil")
+        completion(.failure(NSError(domain: "FirebaseFirestoreSwift",
+                                    code: -1,
+                                    userInfo: [NSLocalizedDescriptionKey:
+                                      "InternalError - Return value and error are both nil"])))
+      }
+    }
+  }
+
+  /// Returns a closure mapped from the a given closure with a `Result` parameter.
+  ///
+  /// - Parameters:
+  ///   - completion: The closure to map.
+  ///   - result: The parameter of the closure to map.
+  /// - Returns: A closure mapped from the given closure.
+  private func mapResultClosure<E>(_ completion: @escaping (_ result: Result<Void, E>) -> Void)
```

Comment:
Does this need to be generic on the error type? If so, why isn't the method above also generic on the error type? If not, prefer the concrete `Error` type.

Reply:
Done